### PR TITLE
Workaround for MSBuild reporting duplicate edges

### PR DESCRIPTION
**Pull request by smera**
MSBuild static graph APIS are sometimes reporting duplicate edges. Temporarily make sure the MSBuild resolver dedups these.

### CODE REVIEW

File: Public/Src/FrontEnd/MsBuild.Serialization/ProjectWithPredictions.cs [hunk 1]
```diff
@@ -48,6 +48,7 @@ public sealed class ProjectWithPredictions<TPathType>
         public PredictedTargetsToExecute PredictedTargetsToExecute { get; }
 
         /// <nodoc/>
+        [JsonIgnore()]
```
Comment (smera) [inline Public/Src/FrontEnd/MsBuild.Serialization/ProjectWithPredictions.cs]: >[JsonIgnore()] [](start = 8, length = 14)

Inconsequential, but without this attribute references are duplicated in the serialized format

File: Public/Src/FrontEnd/MsBuild/PipConstructor.cs [hunk 2]
```diff
@@ -455,6 +455,11 @@ private void SetProcessEnvironmentVariables(IReadOnlyDictionary<string, string>
             processBuilder.AddOutputFile(logDirectory.Combine(PathTable, "msbuild.err"), FileExistence.Optional);
             processBuilder.AddOutputFile(logDirectory.Combine(PathTable, "msbuild.prf"), FileExistence.Optional);
 
+            if (m_resolverSettings?.EnableBinLogTracing == true)
+            {
+                processBuilder.AddOutputFile(logDirectory.Combine(PathTable, "msbuild.binlog"), FileExistence.Optional);
```
Comment (smera) [inline Public/Src/FrontEnd/MsBuild/PipConstructor.cs]: >"msbuild.binlog" [](start = 77, length = 16)

Inconsequential. It is always better to declare a file if we know it is going to be (uniquely) produced. Without this, the containing shared opaque takes ownership.

File: Public/Src/FrontEnd/MsBuild/PipConstructor.cs [hunk 3]
```diff
@@ -455,6 +455,11 @@ private void SetProcessEnvironmentVariables(IReadOnlyDictionary<string, string>
             processBuilder.AddOutputFile(logDirectory.Combine(PathTable, "msbuild.err"), FileExistence.Optional);
             processBuilder.AddOutputFile(logDirectory.Combine(PathTable, "msbuild.prf"), FileExistence.Optional);
 
+            if (m_resolverSettings?.EnableBinLogTracing == true)
```
Comment (jbayardo) [inline Public/Src/FrontEnd/MsBuild/PipConstructor.cs]: Why use the null conditional operator? According to line 438, m_resolverSettings != null

Comment (dannyvv) [reply Public/Src/FrontEnd/MsBuild/PipConstructor.cs]: :+1:

---
In reply to: [273752434](https://github.com/Microsoft/BuildXL/pull/116#discussion_r273752434) [](ancestors = 273752434)

Comment (smera) [reply Public/Src/FrontEnd/MsBuild/PipConstructor.cs]: No real reasons, copy and paste :)

---
In reply to: [273753511](https://github.com/Microsoft/BuildXL/pull/116#discussion_r273753511) [](ancestors = 273753511,273752434)

**Comment by smera** [comment]
> LGTM, though a test that fails when I fix this in msbuild would be nice, to know when to undo this change. It's probably a good idea to keep the bxl graph a 1-to-1 mapping to the msbuild graph as much as possible.

Done!

**Pull request merged**